orttraining/orttraining/training_ops/cuda/activation/gelu_grad_impl_common.cuh

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "orttraining/training_ops/cpu/activation/gelu_computation_mode.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T>
+__device__ __inline__ T ComputeGeluGradScalar(T dY, T X, gelu_computation_mode::Default) {
+  const T kAlpha = T(M_2_SQRTPI) * T(M_SQRT1_2) * T(0.5);
+  return dY * (_Normcdf(X) + X * kAlpha * _Exp(-T(0.5) * X * X));
+}
+
+template <typename T>
+__device__ __inline__ T ComputeGeluGradScalar(T dY, T X, gelu_computation_mode::Approximation) {
+  // copied and adapted from DeepSpeed:
+  // https://github.com/microsoft/DeepSpeed/blob/f5025506de37f617a93eabc2aed7cc4f4bfd7d80/csrc/transformer/gelu_kernels.cu#L10
```

Comment:
>#L10 [](start = 122, length = 4)

nit: do we need line# here? it may be changed any time :-)

Reply:
it shouldn't change - this URL has the exact commit

---
In reply to: [445322351](https://github.com/microsoft/onnxruntime/pull/4146#discussion_r445322351) [](ancestors = 445322351)